library/pkcs7.c

```diff
@@ -546,12 +559,19 @@ int mbedtls_pkcs7_parse_der(mbedtls_pkcs7 *pkcs7, const unsigned char *buf,
     pkcs7->raw.len = buflen;
     end = p + buflen;
 
-    ret = pkcs7_get_content_info_type(&p, end, &pkcs7->content_type_oid);
+    ret = pkcs7_get_content_info_type(&p, end, &end_content_info,
+                                      &pkcs7->content_type_oid);
     if (ret != 0) {
         len = buflen;
         goto try_data;
     }
 
+    /* Ensure PKCS7 data uses the exact number of bytes specified in buflen */
+    if (end_content_info != end) {
+        ret = MBEDTLS_ERR_PKCS7_BAD_INPUT_DATA;
+        goto out;
+    }
+
     if (!MBEDTLS_OID_CMP(MBEDTLS_OID_PKCS7_DATA, &pkcs7->content_type_oid)
         || !MBEDTLS_OID_CMP(MBEDTLS_OID_PKCS7_ENCRYPTED_DATA, &pkcs7->content_type_oid)
```

Comment:
Just spotted, we have `MBEDTLS_OID_PKCS7_ENCRYPTED_DATA` twice in this list. This is harmless, but please correct it. Wondering if it would be clearer to invert the whole condition block and write it as:

```
if (MBEDTLS_OID_CMP(MBEDTLS_OID_PKCS7_SIGNED_DATA, &pkcs7->content_type_oid)) {

Reply:
It indeed would be.

Reply:
I will remove the duplicate entry. But the conditional is important:
```
if (!MBEDTLS_OID_CMP(MBEDTLS_OID_PKCS7_DATA, &pkcs7->content_type_oid)
        || !MBEDTLS_OID_CMP(MBEDTLS_OID_PKCS7_ENVELOPED_DATA, &pkcs7->content_type_oid)
        || !MBEDTLS_OID_CMP(MBEDTLS_OID_PKCS7_SIGNED_AND_ENVELOPED_DATA, &pkcs7->content_type_oid)
        || !MBEDTLS_OID_CMP(MBEDTLS_OID_PKCS7_DIGESTED_DATA, &pkcs7->content_type_oid)
        || !MBEDTLS_OID_CMP(MBEDTLS_OID_PKCS7_ENCRYPTED_DATA, &pkcs7->content_type_oid))
```
returns `MBEDTLS_ERR_PKCS7_FEATURE_UNAVAILABLE` because these are legal oid's but the parsing logic is not implemented.
At the moment, the next conditional is:
```
if (MBEDTLS_OID_CMP(MBEDTLS_OID_PKCS7_SIGNED_DATA, &pkcs7->content_type_oid))
```
Which returns `MBEDTLS_ERR_PKCS7_BAD_INPUT_DATA`, because the oid is not a legal PKCS7 type.

Reply:
The fast path is `MBEDTLS_OID_PKCS7_SIGNED_DATA`, so that should be the outer conditional.

Reply:
True, but I don't think saving a handful of instructions at the start of parsing a PKCS7 file will make a noticeable difference so I'm happy with how it's done in the latest commit.